torchdata/dataloader2/reading_service.py

```diff
@@ -235,3 +240,70 @@ def finalize(self) -> None:
         if self.persistent_workers and self.dl_ is not None and self.dl_._iterator is not None:
             self.dl_._iterator._shutdown_workers()  # type: ignore[attr-defined]
             self.dl_._iterator = None
+
+
+class DistributedReadingService(ReadingServiceInterface):
+    r"""
+    ``DistributedReadingSerivce`` handles distributed sharding on the graph of ``DataPipe`` and
+    guarantee the randomness by sharing the same seed across the distributed processes.
+
+    Args:
+        timeout: Timeout for operations executed against the process group in seconds.
+            Default value equals 30 minutes.
+    """
+
+    def __init__(self, timeout: int = default_timeout_in_s):
+        if not dist.is_available():
+            raise RuntimeError("Torch Distributed is required to be available")
+        self._world_size: int = 1
+        self._rank: int = 0
+        self._datapipe: Optional[DataPipe] = None
+        self._timeout: int = timeout
+        self._pg: Optional[dist.ProcessGroup] = None
+
+    def initialize(self, datapipe: DataPipe) -> DataPipe:
+        r"""
+        Launches the ``gloo``-backend distributed process group. Carries out distributed sharding
+        on the graph of ``DataPipe`` and returnes the graph attached with a ``FullSyncIterDataPipe``
+        at the end.
+        """
+        if not (dist.is_available() and dist.is_initialized()):
+            raise RuntimeError("Torch Distributed is required to be initialized")
+        self._world_size = dist.get_world_size()
+        self._rank = dist.get_rank()
+        self._pg = dist.new_group(backend="gloo", timeout=timedelta(seconds=self._timeout))
+        torch.utils.data.graph_settings.apply_sharding(
+            datapipe,
+            self._world_size,
+            self._rank,
+        )
+        datapipe = datapipe.fullsync(self._timeout)
```

Comment:
super nit: Maybe in future check if `fullsync` already in pipeline (some other cascaded RS added it, or user did it manually).

Reply:
It's a good point. It should be a minimum change with the existing graph functions. Will send a patch shortly.